File: sdks/flutter/lib/src/modules/chat_module.dart
import '../transport/http_client.dart';
import '../transport/websocket_client.dart';
import '../models/message.dart';

class ChatModule {
  final RajutechieStreamKitHttpClient http;
  final RajutechieStreamKitWebSocketClient ws;

  ChatModule({required this.http, required this.ws});

  Future<Map<String, dynamic>> createChannel({
    required String type,
    String? name,
    List<String>? members,
  }) async {
    return http.post('/channels', body: {
      'type': type,
      if (name != null) 'name': name,
      if (members != null) 'members': members,
    });
  }

  Future<Map<String, dynamic>> getChannel(String channelId) async {
    return http.get('/channels/$channelId');
  }

  Future<Map<String, dynamic>> sendMessage(String channelId, {String? text, List<Map<String, dynamic>>? attachments}) async {
    return http.post('/channels/$channelId/messages', body: {
      if (text != null) 'text': text,
      if (attachments != null) 'attachments': attachments,
    });
  }

  Future<Map<String, dynamic>> getMessages(String channelId, {int limit = 25, String? after}) async {
    final params = <String, String>{'limit': '$limit'};
    if (after != null) params['after'] = after;
    return http.get('/channels/$channelId/messages', params: params);
  }

  Stream<Message> messagesStream(String channelId) {
    return ws.on('message.new').map((data) => Message.fromJson(data)).where((msg) => msg.channelId == channelId);
  }

  Future<void> editMessage(String channelId, String messageId, String text) async {
    await http.patch('/channels/$channelId/messages/$messageId', body: {'text': text});
  }

  Future<void> deleteMessage(String channelId, String messageId) async {
    await http.delete('/channels/$channelId/messages/$messageId');
  }

  Future<void> addReaction(String channelId, String messageId, String emoji) async {
    await http.post('/channels/$channelId/messages/$messageId/reactions', body: {'emoji': emoji});
  }

  void startTyping(String channelId) => ws.send('typing.start', {'channelId': channelId});
  void stopTyping(String channelId) => ws.send('typing.stop', {'channelId': channelId});
}

File: sdks/android/streamkit-sdk/src/main/kotlin/com/rajutechie/streamkit/modules/ChatModule.kt
package com.rajutechie.streamkit.modules

import com.rajutechie.streamkit.transport.HttpClient
import com.rajutechie.streamkit.transport.WebSocketClient
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.flow.Flow

data class Channel(val id: String, val type: String, val name: String?, val memberCount: Int)
data class Message(val id: String, val channelId: String, val senderId: String, val content: MessageContent, val createdAt: String)
data class MessageContent(val text: String?, val attachments: List<Attachment>? = null)
data class Attachment(val type: String, val url: String, val filename: String?)
data class ChannelConfig(val type: ChannelType, val name: String?, val members: List<String>?) {
    companion object {
        fun builder() = Builder()
    }
    class Builder {
        private var type: ChannelType = ChannelType.GROUP
        private var name: String? = null
        private var members: List<String>? = null
        fun type(t: ChannelType) = apply { type = t }
        fun name(n: String) = apply { name = n }
        fun members(m: List<String>) = apply { members = m }
        fun build() = ChannelConfig(type, name, members)
    }
}
enum class ChannelType { DIRECT, GROUP, COMMUNITY, OPEN }

class ChatModule(
    private val http: HttpClient,
    private val ws: WebSocketClient,
    private val scope: CoroutineScope
) {
    suspend fun createChannel(config: ChannelConfig): Channel {
        return http.post("/channels", config)
    }

    suspend fun getChannel(channelId: String): Channel {
        return http.get("/channels/$channelId")
    }

    suspend fun sendMessage(channelId: String, text: String, attachments: List<Attachment>? = null): Message {
        val body = mapOf("text" to text, "attachments" to attachments)
        return http.post("/channels/$channelId/messages", body)
    }

    suspend fun getMessages(channelId: String, limit: Int = 25, after: String? = null): List<Message> {
        val params = mutableMapOf("limit" to limit.toString())
        after?.let { params["after"] = it }
        val result: Map<String, Any> = http.get("/channels/$channelId/messages", params)
        @Suppress("UNCHECKED_CAST")
        return result["data"] as? List<Message> ?: emptyList()
    }

    suspend fun editMessage(channelId: String, messageId: String, text: String): Message {
        return http.patch("/channels/$channelId/messages/$messageId", mapOf("text" to text))
    }

    suspend fun deleteMessage(channelId: String, messageId: String) {
        http.delete<Unit>("/channels/$channelId/messages/$messageId")
    }

    suspend fun addReaction(channelId: String, messageId: String, emoji: String) {
        http.post<Unit>("/channels/$channelId/messages/$messageId/reactions", mapOf("emoji" to emoji))
    }

    fun startTyping(channelId: String) {
        ws.emit("typing.start", mapOf("channelId" to channelId))
    }

    fun stopTyping(channelId: String) {
        ws.emit("typing.stop", mapOf("channelId" to channelId))
    }

    fun observeMessages(channelId: String): Flow<Message> {
        return ws.on("message.new", Message::class.java)
    }
}
